In [2]:
import torch
import json
from torch.utils.data import Dataset, DataLoader

In [56]:
# 想把這個做成通用Class, 所以多了幾個額外funcion
class CSI_Dataset(Dataset):
    # 這邊用num不用count當參數, 是怕和str.count()搞混, 所以使用num
    def __init__(self, split='train', mask_list=[], num=-1):
        self.split = split
        with open('./CSI_data.json', 'r') as file:
            self.json_data = json.load(file)

        self.data_split = self.json_data[split]
        
        if len(mask_list):
            # 確定mask list 正確
            # print(f'have mask: {mask_list}')
            data_mask = []
            for mask in mask_list:
                data_mask.extend(self.data_split[index] for index in range(len(self.data_split)) if mask in self.data_split[index])
            self.data = data_mask
        else:
            print(f'No mask: {mask_list}')
            self.data = self.data_split
        
        if num > 0:
            self.__femalecount__(num)
            
    def __len__(self):
        return len(self.data)
            
    def __getitem__(self, idx):
        return self.data[idx]
    
    def __femalecount__(self,num):
        data_count = []
        data_count.extend(name for name in self.data if name.count('F') == num)
        self.data = data_count
                


In [4]:
def Write_Output(output,filename='hw1'):
    data_dict = {i: value for i, value in enumerate(output)}

    with open(f'./{filename}.json', 'w') as json_file:
        json.dump(data_dict, json_file, indent=4) # indent為美化輸出, 很像pretty.print in json output

# Requirement 1

In [13]:
dataset = CSI_Dataset(split='train',mask_list=['Env3'])
dataloader = DataLoader(dataset, batch_size=len(dataset), shuffle=False)
print(dataset.__len__())

have mask: ['Env3']
158110


In [14]:
output = next(iter(dataloader))
output.__len__()

158110

In [15]:
# write the output file
Write_Output(output=output, filename='A1_studentID_studentName_1')

# Requirement 2
In val_set and test_set not contain over 2 female, only consider train set.

In [64]:
# First step to split dataset from train and mash all Female
dataset = CSI_Dataset(split='train',mask_list=['F1','F2','F3'],num=2)

dataloader = DataLoader(dataset, batch_size=len(dataset), shuffle=False)
print(dataset.__len__())

have mask: ['F1', 'F2', 'F3']
257640


In [67]:
output = next(iter(dataloader))
output[247680:257639]

['Env5/npy/F2M3F3/5_posi/240508_231251/1715181209660746998',
 'Env5/npy/F2M3F3/5_posi/240508_231251/1715181209707088810',
 'Env5/npy/F2M3F3/5_posi/240508_231251/1715181209753878886',
 'Env5/npy/F2M3F3/5_posi/240508_231251/1715181209809482756',
 'Env5/npy/F2M3F3/5_posi/240508_231251/1715181209857326770',
 'Env5/npy/F2M3F3/5_posi/240508_231251/1715181209909718340',
 'Env5/npy/F2M3F3/5_posi/240508_231251/1715181209958203564',
 'Env5/npy/F2M3F3/5_posi/240508_231251/1715181210007086441',
 'Env5/npy/F2M3F3/5_posi/240508_231251/1715181210054886767',
 'Env5/npy/F2M3F3/5_posi/240508_231251/1715181210104544798',
 'Env5/npy/F2M3F3/5_posi/240508_231251/1715181210152067894',
 'Env5/npy/F2M3F3/5_posi/240508_231251/1715181210203395159',
 'Env5/npy/F2M3F3/5_posi/240508_231251/1715181210254514197',
 'Env5/npy/F2M3F3/5_posi/240508_231251/1715181210309893716',
 'Env5/npy/F2M3F3/5_posi/240508_231251/1715181210360807778',
 'Env5/npy/F2M3F3/5_posi/240508_231251/1715181210406168995',
 'Env5/npy/F2M3F3/5_posi

In [68]:
# write the output file
Write_Output(output=output, filename='A1_studentID_studentName_2')

# Requirement 3
In this dataset, if only contain 1 female, the format be "Female" not "F", so we just need to mask "Female".<br>
BTW, in val_set and test_set only contain 1 Female or 1 Male, we need to combine val_set, train_set and test_set to record.

In [70]:
dataset_train = CSI_Dataset(split='train', mask_list=['Female'])
dataset_val = CSI_Dataset(split='val', mask_list=['Female'])
dataset_test = CSI_Dataset(split='test', mask_list=['Female'])

print(len(dataset_train), len(dataset_val), len(dataset_test))
dataLoader_train = DataLoader(dataset_train, batch_size=len(dataset_train), shuffle=False)
dataLoader_val = DataLoader(dataset_val, batch_size=len(dataset_val), shuffle=False)
dataLoader_test = DataLoader(dataset_test, batch_size=len(dataset_test), shuffle=False)

have mask: ['Female']
have mask: ['Female']
have mask: ['Female']
229679 3622 4807


In [77]:
# combine all dataset to output, and check it contain all of "Female" 
output = []
output.extend(next(iter(dataLoader_train)))
output.extend(next(iter(dataLoader_val)))
output.extend(next(iter(dataLoader_test)))
print(len(output))
output[233001:233301]

238108


['val_set/npy/Female3/rand_posi/240509_111828/1715224764352084857',
 'val_set/npy/Female3/rand_posi/240509_111828/1715224764397858836',
 'val_set/npy/Female3/rand_posi/240509_111828/1715224764444951954',
 'val_set/npy/Female3/rand_posi/240509_111828/1715224764492160867',
 'val_set/npy/Female3/rand_posi/240509_111828/1715224764537815576',
 'val_set/npy/Female3/rand_posi/240509_111828/1715224764584097742',
 'val_set/npy/Female3/rand_posi/240509_111828/1715224764630081099',
 'val_set/npy/Female3/rand_posi/240509_111828/1715224764677552059',
 'val_set/npy/Female3/rand_posi/240509_111828/1715224764724263855',
 'val_set/npy/Female3/rand_posi/240509_111828/1715224764771677353',
 'val_set/npy/Female3/rand_posi/240509_111828/1715224764819196479',
 'val_set/npy/Female3/rand_posi/240509_111828/1715224764866966869',
 'val_set/npy/Female3/rand_posi/240509_111828/1715224764914710796',
 'val_set/npy/Female3/rand_posi/240509_111828/1715224764962805620',
 'val_set/npy/Female3/rand_posi/240509_111828/17

In [78]:
# write the output file
Write_Output(output=output, filename='A1_studentID_studentName_3')

# Requriement 4
In this case, I want to use mask for y/m/d time, and to compare the h/m/s time.<br>
In val_set and test_set, time start at 240509, so i ignore val_set and test_set.<br>
Hit: 240509 isn't only in time format, also in filename. 